content/posts/2025-02-13-perf-improvements/index.md

```diff
@@ -0,0 +1,237 @@
+
+```
+Summary:
+  Total:	30.0017 secs
+  Slowest:	0.1207 secs
+  Fastest:	0.0006 secs
+  Average:	0.0015 secs
+  Requests/sec:	39863.3677
+  
+  Total data:	735520320 bytes
+  Size/request:	735 bytes
+
+Response time histogram:
+  0.001 [1]	|
+  0.013 [999261]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
+  0.025 [538]	|
+  0.037 [70]	|
+  0.049 [32]	|
+  0.061 [38]	|
+  0.073 [9]	|
+  0.085 [1]	|
+  0.097 [0]	|
+  0.109 [0]	|
+  0.121 [50]	|
+
+Latency distribution:
+  10% in 0.0008 secs
+  25% in 0.0009 secs
+  50% in 0.0011 secs
+  75% in 0.0014 secs
+  90% in 0.0018 secs
+  95% in 0.0023 secs
+  99% in 0.0038 secs
+
+Details (average, fastest, slowest):
+  DNS+dialup:	0.0000 secs, 0.0000 secs, 0.0044 secs
+  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0048 secs
+  req write:	0.0000 secs, 0.0000 secs, 0.0082 secs
+  resp wait:	0.0014 secs, 0.0006 secs, 0.1193 secs
+  resp read:	0.0001 secs, 0.0000 secs, 0.0223 secs
+
+Status code distribution:
+  [200]	1000000 responses
+```
+
+After that I ran the same command twice, first with NetObserv 1.7, second with NetObserv 1.8.
+
+In both cases, some values remained unchanged compared to the baseline: Fastest was always 0.0006s, Average 0.0015s, and same for the Fastest and Average details (DNS+dialup, DNS-lookup, req write, resp wait, resp read). These numbers are a good sign that NetObserv doesn't generate latency overhead for most of the requests. It's on higher percentiles that an overhead starts showing up.
+
+**Fig. 5: latency distribution across baseline and versions (in seconds)**
+
+![Latency distribution](./latency-distribution.png)
+
+Unlike in 1.7, at p50 1.8 shows no latency overhead compared to the baseline, meaning that half of the requests have pretty much no overhead. At p99, the latency overhead in 1.7 was +1 millisecond, in 1.8 it decreased to +0.7 milliseconds.
+
+It's also in terms of maximum queries per second that the overhead is interesting to see.
+
+**Fig. 6: QPS across baseline and versions**
+
+![QPS](./qps.png)
+
+Since this is not rate-limited, _hey_ generates as much load as it can with its 50 workers. It shows a QPS increase of +5.6% compared to NetObserv 1.7.
+
+## How did we go there: under the cover
+
+Now is the technical part. Here's what we did.
+
+### Moving away from per-CPU map
+
+eBPF provides several [structures](https://docs.ebpf.io/linux/map-type/) that allow sharing data between the kernel space and the user space. Hash maps are especially useful as they allow for data persistence across calls and hook points. Thanks to hash maps, NetObserv can keep track of flows in the kernel space, aggregate new packets observed there, and thus limit the size of data transfer between the kernel and the user spaces for better performance.
+
+Among the different map types, there is [BPF_MAP_TYPE_HASH](https://docs.ebpf.io/linux/map-type/BPF_MAP_TYPE_HASH/) and [BPF_MAP_TYPE_PERCPU_HASH](https://docs.ebpf.io/linux/map-type/BPF_MAP_TYPE_PERCPU_HASH/).
+
```

Comment:
I know this PR is already merged, but I have a another comment now.

The diagram for Per-CPU hash map shows a single hash map, but in fact, according to [BPF_MAP_TYPE_PERCPU_HASH](https://docs.ebpf.io/linux/map-type/BPF_MAP_TYPE_PERCPU_HASH/), it is actually "a separate hash map for each logical CPU".  It should look like the diagram on the left, but one for each CPU.  Retrieving a value is the same regardless of which model you choose, because  you are given the right hash map to begin with when you are in a function, whereas the Per-CPU diagram seems to imply that the value is a list.

Reply:
I am not exactly sure how the internal implementation is, but I can tell that when per-cpu map are fetched from the user space, it's a list that you get: see the diff here : https://github.com/netobserv/netobserv-ebpf-agent/pull/469/files#diff-daa370b7c53b55b200b08ed318bef1a61e5838f5f5c5ded3442c83caf4e96f54L726
Previously we fetched `map[ebpf.BpfFlowId][]ebpf.BpfFlowMetrics` (so, a list of `BpfFlowMetrics` - one per CPU)  , and now without per-cpu map we have `map[ebpf.BpfFlowId]model.BpfFlowContent`, so no list.

Reply:
but maybe my representation here is only valid as from a user-space perspective, using the provided bpf helpers , with some transformation happening under the hood